WORKFLOW.md

```diff
@@ -0,0 +1,50 @@
+_below is more like open disscussion than a design doc_
+### Introduction
+
+The following CRDs are defined for managing the lifecycle of Buckets:
+
+ - BucketRequest - Represents a request to provision a Bucket
+ - BucketClass - Represents a class of Buckets with similar characteristics
+ - Bucket - Represents a Bucket or its equivalent in the storage backend
+
+ The following CRDs are defined for managing the lifecycle of workloads accessing the Bucket:
+
+ - BucketAccessRequest - Represents a request to access a Bucket
+ - BucketAccessClass - Represents a class of accessors with similar access requirements
+ - BucketAccess - Represents a access token or service account in the storage backend
+
+Simialr to CSI, it has three running components :
+- Controller - mananges life cycle of COSI objects, watches for the requests for above mentioned objects
+- Node adapter - bootstraps bucket access request to applications for their worloads
+- Sidecar/Driver - receives the request and calls corresponding api in the driver
+
+More details and defintions can be found the [COSI Spec](https://github.com/kubernetes-sigs/container-object-storage-interface-spec/blob/master/spec.md)
+
+### Workflow
+
+- BucketRequest(namespace scoped) can be created with reference to a BucketClass, so user will get a Bucket(clusterscoped). The COSI driver will create a `backend bucket` in RGW for a Bucket Request
+- BucketAccessRequest(namespace scoped) can be create with reference to a BucketAccessClass and BucketRequest, so user will get a BucketAccess(clusterscoped).The COSI driver will create a `s3 user` with proper access
```

Comment:
we should also support the case where we use an exiting S3 use. we can have deployments with a very large number of buckets, and we probably don't want to create as many users.

Reply:
COSI does support this "brownfield" use-case by allowing admins to create a COSI "Bucket" manually. We should make sure we understand what COSI enforces onto its drivers and what options we have within Rook for implementation. I don't think it's a bad decision to create a new user for each access, as this could allow for easier management around who has access if someone's access needs to be revoked.

Reply:
my comment is not only for the brownfield case. tying the number of users to the number of bucket accesses may pose a problem to the RGW. the mechanism for support a very large number of buckets may not be present in the case of users.

Reply:
Is there any known limit for users can be accessed per bucket?
 Currently as per my understanding, `BucketRequest` is an abstraction over backend Bucket and There can be multiple `BucketAccessRequest` referring to a `Bucket`(CRD) created by `BucketRequest` and applications in k8s will assign `BucketAccess`(should refer to backend user IMO)  @jeffvance : Any more comments?

Reply:
@thotz a COSI cluster-scoped `Bucket` instance is an abstraction of a backend bucket. A `BucketRequst` (BR) is a user request to create a new (greenfield) backend bucket. A `BucketAccessRequest` (BAR) is a user request to access an existing backend bucket, and the cluster-scoped `BucketAccess` (BA) instance is an abstraction of this access. Multiple BAs can reference the same `Bucket` instance, and multiple BARs can reference the same BR as long as they are in the namespace of the BR.

Reply:
so, going back to the discussion of users and bucket.
* creating a Bucket, is just a definition, and does not need to create an actual bucket in the RGW
* creating a BucketRequest would need to include an S3 user name (maybe as part of the opaque parameter map) so that not all buckets are created by some hardcoded user
* creating a BucketAccessRequest creates an S3 user and grants them access to this bucket. Note that the BAR also has map of opaque data, so, in theory, if an S3 user is provided there, a new user does not need to be created for each BAR, and an existing user may be reused. in this case, the driver may fetch the access and secret keys via `radosgw-admin` and return them in the reply

Reply:
> creating a Bucket, is just a definition, and does not need to create an actual bucket in the RGW

COSI expects the driver to attempt to create a backend bucket and, if successful, return its unique bucket ID. Note: COSI also expects the driver to be idempotent such that if the backend bucket already exists, the driver returns a status indicating that the bucket exists and did not need to be created.

> creating a BucketRequest would need to include an S3 user name (maybe as part of the opaque parameter map) so that not all buckets are created by some hardcoded user

Not correct. COSI created buckets are not owned by the BR. They are owned by the driver and when the BAR is processed the driver grants access to that request. Also note that a BR does not have a map.  I'm not sure where you've seen this but please send me a link so we can correct it. 

> creating a BucketAccessRequest creates an S3 user and grants them access to this bucket. Note that the BAR also has map of opaque data, so, in theory, if an S3 user is provided there, a new user does not need to be created for each BAR, and an existing user may be reused.

The BAR does _not_ have a map (same comment as above for the BR). The BAR references a `BucketAccessClass` (BAC) and optionally a ServiceAccount. If the SA name is omitted then the driver will most likely mint a new user.

Reply:
@jeffvance thanks for clarifying!

> COSI expects the driver to attempt to create a backend bucket and, if successful, return its unique bucket ID.
> ... Not correct. COSI created buckets are not owned by the BR. They are owned by the driver

This means that the admin user is the owner to all of the buckets?

> Note: COSI also expects the driver to be idempotent such that if the backend bucket already exists, the driver returns a status indicating that the bucket exists and did not need to be created.

According to [this](https://docs.ceph.com/en/latest/radosgw/s3/bucketops/#put-bucket) the operation is idempotent, but does not give an indication that the bucket already exists.

>  I'm not sure where you've seen this but please send me a link so we can correct it.

OK, I hope I found the right place now. Just want to verify it is [here](https://github.com/kubernetes-sigs/container-object-storage-interface-api)?

> The BAR does not have a map (same comment as above for the BR). The BAR references a BucketAccessClass (BAC) and optionally a ServiceAccount. If the SA name is omitted then the driver will most likely mint a new user.

OK. So, this would probably be done by adding bucket policies by the admin user.

Reply:
I think Casey will write more, but this does seem problematic.  In AWS, buckets are owned by a controlling account (tenant), and this is a basic abstraction COSI would need to support a service bureau style of operation.  RGW has operated at a multi-tenant scale for years.

Reply:
(I'm worried that we need to have a discussion around this in the COSI upstream.)

Reply:
> @jeffvance thanks for clarifying!
> 
> > COSI expects the driver to attempt to create a backend bucket and, if successful, return its unique bucket ID.
> > ... Not correct. COSI created buckets are not owned by the BR. They are owned by the driver
> 
> This means that the admin user is the owner to all of the buckets?

* rgw defaults to a limit of 1000 buckets per user, and has some known scaling issues past that
* for purposes of quota, all objects in a bucket are attributed to the bucket's owner. so this strategy means we can't offer meaningful per-user quota enforcement

Reply:
@jeffvance did a nice write-up for the cosi workflow which help to understand bit more about COSI https://docs.google.com/document/d/1-XwQ3ZbGU_vqsl2rwQQIg1hSEvvjNaOqlqiRFGEuqQo/edit#

Reply:
> @jeffvance thanks for clarifying!
> 
> > COSI expects the driver to attempt to create a backend bucket and, if successful, return its unique bucket ID.
> > ... Not correct. COSI created buckets are not owned by the BR. They are owned by the driver
> 
> This means that the admin user is the owner to all of the buckets?


I had this discussion with COSI team regarding the ownership. So I can answer this question, above is one way to do it. Another approach is set to make relation with BucketClass. There is will an owner for all the Buckets(Greenfield case)  referring to that BucketClass. This owner will be created upon first BucketRequest. We can also details about this owner in the `Opaque` field of BucketClass() if needed. And one more thing the opaque field will only be available in `BucketClass` or `BucketAccessClass` not in `BucketRequest` or `BucketAccessRequest` due to portability issue(using same BR/BAR when app is ported to different cluster/enviroment). I don't fully understand reason for the same 

> 
> > Note: COSI also expects the driver to be idempotent such that if the backend bucket already exists, the driver returns a status indicating that the bucket exists and did not need to be created.
> 
> According to [this](https://docs.ceph.com/en/latest/radosgw/s3/bucketops/#put-bucket) the operation is idempotent, but does not give an indication that the bucket already exists.
> 
> > I'm not sure where you've seen this but please send me a link so we can correct it.
> 
> OK, I hope I found the right place now. Just want to verify it is [here](https://github.com/kubernetes-sigs/container-object-storage-interface-api)?
> 
> > The BAR does not have a map (same comment as above for the BR). The BAR references a BucketAccessClass (BAC) and optionally a ServiceAccount. If the SA name is omitted then the driver will most likely mint a new user.
> 
> OK. So, this would probably be done by adding bucket policies by the admin user.